vfs: add option to read source files in chunks

B4dM4n opened the pull request:
This is PR  adds the option `--vfs-read-chunk-size`to the vfs package.
The intention behind this is to reduce the used download quota for some combinations of remotes and workloads.
In my case it's a Google Drive remote containing large files which are only read in small parts. It seems like Drive is counting every requested byte to the daily quota even when the file is not fully downloaded. Using `Range` requests the number of requested bytes can be reduced at the cost of an increased number of requests.

For the tests to keep working i needed to implement the `fs.RangeOption` for the local remote (i can split that into a separate commit if desired)

ncw commented on vfs/read.go:
```diff
@@ -62,7 +65,11 @@ func (fh *ReadFileHandle) openPending() (err error) {
-	r, err := fh.o.Open()
+	var end int64 = -1
+	if fh.chunkSize > 0 {
+		end = fh.chunkSize - 1
```
This should probably cap end to o.Size(), or maybe add another condition in the if `fh.chunkSize > 0 && fh.chunkSize < o.Size()`?

ncw commented on vfs/read.go:
```diff
@@ -62,7 +65,11 @@ func (fh *ReadFileHandle) openPending() (err error) {
-	r, err := fh.o.Open()
+	var end int64 = -1
+	if fh.chunkSize > 0 {
+		end = fh.chunkSize - 1
```
This should probably cap end to o.Size(), or maybe add another condition in the if fh.chunkSize > 0 && fh.chunkSize < o.Size()?

B4dM4n replied:
The check should not be necessary. The HTTP RFC [https://tools.ietf.org/html/rfc7233#section-2.1](7233#2.1) specifies: 
>A client can limit the number of bytes requested without knowing the
> size of the selected representation.  If the last-byte-pos value is
> absent, or if the value is greater than or equal to the current
> length of the representation data, the byte range is interpreted as
> the remainder of the representation (i.e., the server replaces the
> value of last-byte-pos with a value that is one less than the current
> length of the selected representation).

In short, the HTTP server has to correct the end value to the correct position.
My tests using Google Drive confirmed this behavior.

I can add the check to reduce the chance of a broken HTTP server rejecting requests beyond the real end.

ncw replied:
I think RangeOption should follow the RFC here, so your code is good as it stands.  I added some extra docs to the help in 84bc4dc142424ad0bc9f9896939e5a7c09078a5c and as part of #1825 I'll make sure this gets integration tested.

ncw commented on vfs/read.go:
```diff
@@ -120,12 +127,17 @@ func (fh *ReadFileHandle) seek(offset int64, reopen bool) (err error) {
 		}
+		var end int64 = -1
+		if fh.chunkSize > 0 {
+			end = offset + fh.chunkSize - 1
```
This might need capping to o.Size() too...

ncw commented on vfs/read.go:
```diff
@@ -225,10 +237,22 @@ func (fh *ReadFileHandle) readAt(p []byte, off int64) (n int, err error) {
+					doReopen = true
+					off = newOffset
+					p = p[n:]
+					fh.chunkSize *= 2
```
Why do you think doubling the chunk size each time is a good idea?  Maybe it should have a maximum?

B4dM4n replied:
The intention is to keep the number of requests small. A file being read for many chunks continuously is probably being read to further.
Most providers don't provide information, if the amount of "downloaded bytes" or the "requested bytes" are limited.
For Google Drive i read a value of 10TB downloads per day. From my experience they seem to count the requested bytes, because i could download parts of large files, after the whole file download was returning "403 Access Denied".
Probably most providers count "requested bytes", but there could be other limits as well.

Maybe a second option `--vfs-read-chunk-growth` could give this decision to the user, where the default is to keep the chunk size and with the option enabled chunk size doubles.

ncw replied:
Requested bytes are the easy thing to measure for the cloud provider.  Downloaded bytes are tricker to measure...

An option for the user to control the doubling sounds like a good idea to trade latency for efficiency if required.

Do you think the buffer size should be reset to the default on seeks too?

B4dM4n replied:
Resetting the size on seek sounds good.I will add that.

ncw commented:
I think this is a very interesting idea and it has worried me that the open is effectively requesting the whole file when we only read part of it...  I think the chunked reading should probably be the default.

Can you split the changes to the local into a separate PR and market it with #1825

If I merge your code as-is then it will break the VFS layer for remotes which don't support a ranged open.  So either we need to complete #1825 first or put a fallback in your code to use the Seek open if the chunks aren't specified.

What do you think?

B4dM4n commented:
I will spit the local changes in a second PR.

You are right, i'm currently assuming `Range` support for every remote, which is very bad (i even ignore the no-seek option).

What about a `fs.Features` like `RangeSupported`. This would allow each remote to specify the support at runtime.
The http remote could get a config file option `range-supported = false` to disable the feature on unsupported servers.

ncw commented:
> What about a fs.Features like RangeSupported. This would allow each remote to specify the support at runtime.
> The http remote could get a config file option range-supported = false to disable the feature on unsupported servers.

My plan in #1825 was to add RangeOption to all the backends as that will be the most useful the rest of rclone.

I hadn't thought of http remotes with range not supported though... I'm not sure it will be a big problem in practice though.

B4dM4n commented:
Bad HTTP servers not supporting Range request shouldn't exist anymore.

I think we should continue the conversation about range support in #1825 as the related issue.

B4dM4n commented:
I updated this PR to make use of the recently added `ChunkedReader`. For this purpose extended the `ChunkedReader` to make use of `RangeSeek` of the wrapped reader and replaced the chunk growth parameter with a chunk size limit.

There are two new flags for `mount` now:
`--vfs-read-chunk-size` which enables the chunked reading when greater than zero and
`--vfs-read-chunk-size-limit` to enable growing of the chunk size. When `--vfs-read-chunk-size-limit` is greater than `--vfs-read-chunk-size`, the chunk size will get doubled after each chunk read until the limit is reached. The chunk size will be reset to `--vfs-read-chunk-size` after each `Seek`.

ncw commented on cmd/mountlib/mount.go:
```diff
@@ -154,6 +154,21 @@ to use Type=notify. In this case the service will enter the started state
+
+--vfs-read-chunk-size will enable reading the source objects in parts.
+This can reduce the used download quota for some remotes by requesting only chunks
+from the remote that are actually read at the cost of an increased number of requests.
```
You should probably mention here that it only works for --vfs-cache-mode < full

B4dM4n replied:
This is a good idea.
I added notice to the docs in the latest commit.

ncw commented on fs/chunkedreader/chunkedreader.go:
```diff
@@ -1,12 +1,18 @@
 package chunkedreader
```
chunkedreader could do with a test!  Perphaps in a later commit?

B4dM4n replied:
I added some tests in the latest version.
They are not testing all the features `ChunkedReader` has, but should ensure that reading and seeking yield correct results.

ncw commented:
Do you want to get this in for v1.40?

B4dM4n commented:
Sorry for the delay.
I finally got the time to write some tests and fix bugs in the early version of `ChunkedReader`.

I also added the option to set `--vfs-read-chunk-size-limit off` (or `-1`) to disable the limit, meaning infinite growth.

It should be ready to be tested in beta now. By default the feature is still disabled.

ncw commented:
Thanks for the tests.  I've given the code another once over and I think it looks good!

We can run it in the beta and people can experiment with the flag defaults.

Thank you very much :-)
